### PR TITLE
ldelf: add support for relocation type R_*_NONE

### PR DESCRIPTION
**Pull request by etienne-lms**

### CODE REVIEW

File: ldelf/ta_elf_rel.c [hunk 1]
```diff
@@ -327,6 +327,9 @@ static void e32_relocate(struct ta_elf *elf, unsigned int rel_sidx)
 		size_t sym_idx = 0;
 		vaddr_t val = 0;
 
+		if (ELF32_R_TYPE(rel->r_info) == R_ARM_NONE)
+			continue;
```
Comment (etienne-lms) [inline ldelf/ta_elf_rel.c:331]: Maybe I was a bit rude testing this here. I cannot find back where I saw the offset value was not to consider, at least for 32b and 64b Arm architecture.

Maybe a better fix would be to move this test in the switch/case below:
```patch
 		switch (ELF32_R_TYPE(rel->r_info)) {
+		case R_ARM_NONE:
+			break;
 		case R_ARM_ABS32:
```

**Review by jforissier — COMMENTED** [review]
The presence of an R_ARM_NONE relocation in the TA is weird, because it is a static relocation and therefore it is not supposed to be interpreted by the dynamic loader. The "ELF for the ARM Architecture" specification says:
> R_ARM_NONE records that the section containing the place to be relocated depends on the section defining the symbol mentioned in the relocation directive in a way otherwise invisible to the static linker. The effect is to prevent removal of sections that might otherwise appear to be unused.

Why the linker would leave it in the executable after the final link phase is a mystery to me...
Are you using a special linker script or linker options? I'm just wondering if this could indicate a mistake that could cause other issues later.

**Comment by etienne-lms** [comment]
> Are you using a special linker script or linker options? I'm just wondering if this could indicate a mistake that could cause other issues later.

I built platform stm32mp1 using a pristine 3.10.0... but with some hacks in **build**:
* rename the toolchain tools basename: s/arm-linux-gnueabihf/arm-none-linux-gnueabihf/g
* upgrading with `BR2_TOOLCHAIN_EXTERNAL_GCC_9=y`
* also upgrading with `BR2_TOOLCHAIN_EXTERNAL_HEADERS_5.4=y`

and redirecting **toolchains/aarch32/** to prebuilt Arm GCC-9.2 tree downloaded from [here](https://developer.arm.com/tools-and-software/open-source-software/developer-tools/gnu-toolchain/gnu-a/downloads).

The same can be done with the platform qemu but that **buildroot** needs to be upgraded to. I used 2020.02 and removed the dependency on `host-python-pycrypto` that fails with this buildroot release tag.

File: ldelf/ta_elf_rel.c (hunk 1 above)
Comment (etienne-lms) [reply ldelf/ta_elf_rel.c:331]: ok, i'll fix.

**Comment by jenswi-linaro** [comment]
> The presence of an R_ARM_NONE relocation in the TA is weird, because it is a static relocation and therefore it is not supposed to be interpreted by the dynamic loader. The "ELF for the ARM Architecture" specification says:
> 
> > R_ARM_NONE records that the section containing the place to be relocated depends on the section defining the symbol mentioned in the relocation directive in a way otherwise invisible to the static linker. The effect is to prevent removal of sections that might otherwise appear to be unused.
> 
> Why the linker would leave it in the executable after the final link phase is a mystery to me...

Yeah, this sounds like a broken toolchain. Perhaps there's some compiler options that doesn't play well...

**Comment by etienne-lms** [comment]
Updated.

So this fix seems more a workaround. Should it be stated (that such relocation type are not expected) in the commit log or an a debug trace?

**Comment by jenswi-linaro** [comment]
> So this fix seems more a workaround. Should it be stated (that such relocation type are not expected) in the commit log or an a debug trace?

Mentioning it in the commit message and perhaps a short comment in the code should be enough. Logging seems a bit much.

**Comment by etienne-lms** [comment]
Updated with a fixup commit for the added inline comments.

**Comment by jenswi-linaro** [comment]
`Reviewed-by: Jens Wiklander <jens.wiklander@linaro.org>`

**Comment by jforissier** [comment]
Look good to me too.

`Reviewed-by: Jerome Forissier <jerome@forissier.org>`

**Comment by etienne-lms** [comment]
Squashed (+ minor typo fix in inline comment) and tags applied.

**Pull request merged**